src/CrsGraph.C

```diff
@@ -0,0 +1,1031 @@
+    const bool throwIfMasterNotFound = false;
+    for(const stk::mesh::Bucket* bptr : nodeBuckets) {
+        const stk::mesh::Bucket& b = *bptr;
+        const stk::mesh::EntityId* nodeIds = stk::mesh::field_data(*realm_.naluGlobalId_, b);
+        for(size_t i=0; i<b.size(); ++i) {
+            stk::mesh::Entity node = b[i];
+
+            GlobalOrdinal gid =-1;
+            // needed because of some shared and ghost nodes. 
+            if (nodeIds[i] != bulk.identifier(node)) {
+              stk::mesh::Entity master = get_entity_master(bulk, node, nodeIds[i],
+                                                           throwIfMasterNotFound);
+              if (bulk.is_valid(master) && master != node) {
+                gid = * stk::mesh::field_data(*realm_.tpetGlobalId_, master);
+                entityToColLID_[node.local_offset()] = totalColsMap_->getLocalElement(gid);
+              }
+              else if (!bulk.is_valid(master)) {
+                gid = -1;
+                entityToColLID_[node.local_offset()] = gid;
+              }
+              else {
+                gid =  * stk::mesh::field_data(*realm_.tpetGlobalId_, node);
+              }
+            }
+            else
+              gid =  * stk::mesh::field_data(*realm_.tpetGlobalId_, node);
+
+            if(gid == 0 || gid == -1 || gid == std::numeric_limits<LinSys::GlobalOrdinal>::max() ) {
+	      // unit_test1 does produce ghost nodes that have a master that don't have a valid tpetGlobalId_
+              entityToColLID_[node.local_offset()] = -1;
+            }
+            else
+              entityToColLID_[node.local_offset()] = totalColsMap_->getLocalElement(gid);
+        }
+    }
+} //fill_entity_to_col_LID_mapping
+
+void CrsGraph::storeOwnersForShared()
+{
+  const stk::mesh::BulkData & bulkData = realm_.bulk_data();
+  const stk::mesh::MetaData & metaData = realm_.meta_data();
+  const stk::mesh::Selector all = metaData.universal_part() & !(realm_.get_inactive_selector());
+  const stk::mesh::BucketVector& buckets = realm_.get_buckets( stk::topology::NODE_RANK, all );
+
+  for(const stk::mesh::Bucket* bptr : buckets) {
+    const stk::mesh::Bucket& bkt = *bptr;
+    for(stk::mesh::Entity node : bkt) {
+      int status = getDofStatus(node);
+      if (status & DS_SharedNotOwnedDOF) {
+        stk::mesh::EntityId naluId = *stk::mesh::field_data(*realm_.naluGlobalId_, node);
+        stk::mesh::Entity master = get_entity_master(bulkData, node, naluId);
+        GlobalOrdinal gidbase = * stk::mesh::field_data(*realm_.tpetGlobalId_, master);
+        ThrowRequire(gidbase != 0);
+        for(unsigned idof=0; idof < numDof_; ++ idof) {
+          GlobalOrdinal gid = gidbase+idof;
+          ownersAndGids_.insert(std::make_pair(bulkData.parallel_owner_rank(master), gid));
+        }
+      }
+    }
+  }
+} //storeOwnersForShared
+
+void CrsGraph::finalizeGraph()
+{
+  ThrowRequire(inConstruction_);
```

Comment:
We probably want to avoid "re-finalizing" the same graph.  Combined with the issue with the tpetra global id potentially being in a bad state I mentioned before, finalizations after the first call may be incorrect depending on the order in which the equation systems are sequenced.